File: desktop/windows/src/renderer/src/lib/modelConfig.ts
// User's model/provider selection + BYOK credentials, persisted in localStorage.
//
// Cortex routes its agentic LLM calls either through the bundled Cortex backend
// (default — works out of the box, no setup) or, once the user configures one
// here, directly through a chosen local/cloud provider (see shared/providers.ts).
import { getProvider, type ProviderInfo, type ModelInfo } from '../../../shared/providers'

const KEY = 'cortex-model-config-v1'

export type ModelConfig = {
  /** 'backend' = Cortex proxy (default). 'provider' = use the selected provider below. */
  mode: 'backend' | 'provider'
  /** Selected provider id from shared/providers.ts (when mode === 'provider'). */
  providerId?: string
  /** Selected model id within that provider. */
  modelId?: string
  /** Per-provider API keys (BYOK). Keyed by provider id. Stored locally only. */
  apiKeys: Record<string, string>
  /** Override base URL (for the "custom" provider, or self-hosted endpoints). */
  customBaseUrl?: string
}

const defaults: ModelConfig = { mode: 'backend', apiKeys: {} }

function load(): ModelConfig {
  try {
    const raw = localStorage.getItem(KEY)
    if (!raw) return { ...defaults }
    const parsed = JSON.parse(raw) as Partial<ModelConfig>
    return { ...defaults, ...parsed, apiKeys: { ...(parsed.apiKeys ?? {}) } }
  } catch {
    return { ...defaults }
  }
}

let current: ModelConfig = load()
const listeners = new Set<(c: ModelConfig) => void>()

function persist(): void {
  try {
    localStorage.setItem(KEY, JSON.stringify(current))
  } catch {
    /* quota / privacy mode */
  }
  listeners.forEach((cb) => cb(current))
}

export function getModelConfig(): ModelConfig {
  return current
}

export function setModelConfig(patch: Partial<ModelConfig>): void {
  current = { ...current, ...patch, apiKeys: { ...current.apiKeys, ...(patch.apiKeys ?? {}) } }
  persist()
}

export function setApiKey(providerId: string, key: string): void {
  current = { ...current, apiKeys: { ...current.apiKeys, [providerId]: key } }
  persist()
}

export function getApiKey(providerId: string): string {
  return current.apiKeys[providerId] ?? ''
}

export function onModelConfigChange(cb: (c: ModelConfig) => void): () => void {
  listeners.add(cb)
  return () => listeners.delete(cb)
}

export type ResolvedTarget = {
  provider: ProviderInfo
  model: ModelInfo | undefined
  baseUrl: string
  apiKey: string
}

/**
 * Resolve the active provider target, or null when Cortex should use the default
 * backend (mode === 'backend', nothing configured, or a missing required key).
 */
export function resolveTarget(): ResolvedTarget | null {
  if (current.mode !== 'provider' || !current.providerId) return null
  const provider = getProvider(current.providerId)
  if (!provider) return null
  const baseUrl =
    (provider.id === 'custom' ? current.customBaseUrl : provider.baseUrl)?.trim() || ''
  if (!baseUrl) return null
  const apiKey = getApiKey(provider.id)
  if (provider.requiresApiKey && !apiKey) return null
  const model = provider.models.find((m) => m.id === current.modelId)
  return { provider, model, baseUrl, apiKey }
}

File: app/lib/services/cortex/model_config.dart
// Cortex model/provider selection + BYOK credentials + cloud-sync stub.
// Persisted via SharedPreferencesUtil. Flutter port of modelConfig.ts / cloudSync.ts.
import 'dart:convert';

import 'package:omi/backend/preferences.dart';
import 'package:omi/services/cortex/license.dart';
import 'package:omi/services/cortex/providers.dart';

enum CortexEngineMode { backend, provider }

class ResolvedTarget {
  final ProviderInfo provider;
  final ModelInfo? model;
  final String baseUrl;
  final String apiKey;
  const ResolvedTarget(this.provider, this.model, this.baseUrl, this.apiKey);
}

class CortexModelConfig {
  CortexModelConfig._();
  static final CortexModelConfig instance = CortexModelConfig._();

  final _prefs = SharedPreferencesUtil();

  CortexEngineMode get mode => _prefs.getString('cortexEngineMode', defaultValue: 'backend') == 'provider'
      ? CortexEngineMode.provider
      : CortexEngineMode.backend;
  set mode(CortexEngineMode v) => _prefs.saveString('cortexEngineMode', v == CortexEngineMode.provider ? 'provider' : 'backend');

  String? get providerId {
    final v = _prefs.getString('cortexProviderId');
    return v.isEmpty ? null : v;
  }

  set providerId(String? v) => _prefs.saveString('cortexProviderId', v ?? '');

  String? get modelId {
    final v = _prefs.getString('cortexModelId');
    return v.isEmpty ? null : v;
  }

  set modelId(String? v) => _prefs.saveString('cortexModelId', v ?? '');

  String? get customBaseUrl {
    final v = _prefs.getString('cortexCustomBaseUrl');
    return v.isEmpty ? null : v;
  }

  set customBaseUrl(String? v) => _prefs.saveString('cortexCustomBaseUrl', v ?? '');

  Map<String, String> get _apiKeys {
    try {
      final raw = _prefs.getString('cortexApiKeys');
      if (raw.isEmpty) return {};
      return Map<String, String>.from(jsonDecode(raw) as Map);
    } catch (_) {
      return {};
    }
  }

  String apiKey(String providerId) => _apiKeys[providerId] ?? '';

  void setApiKey(String providerId, String key) {
    final keys = _apiKeys;
    keys[providerId] = key;
    _prefs.saveString('cortexApiKeys', jsonEncode(keys));
  }

  /// Resolve the active provider target, or null when Cortex should use the
  /// default backend (mode backend, nothing configured, or missing required key).
  ResolvedTarget? resolveTarget() {
    if (mode != CortexEngineMode.provider || providerId == null) return null;
    final provider = cortexProviderById(providerId);
    if (provider == null) return null;
    final base = (provider.id == 'custom' ? (customBaseUrl ?? '') : provider.baseUrl).trim();
    if (base.isEmpty) return null;
    final key = apiKey(provider.id);
    if (provider.requiresApiKey && key.isEmpty) return null;
    return ResolvedTarget(provider, cortexModel(provider.id, modelId), base, key);
  }
}

/// Cloud sync — Pro feature, stub. Records intent + a local marker; the real
/// implementation will push an encrypted snapshot to cortex.apym.io.
class CortexCloudSync {
  CortexCloudSync._();
  static final CortexCloudSync instance = CortexCloudSync._();

  final _prefs = SharedPreferencesUtil();
  static const String endpoint = 'https://cortex.apym.io/api/sync';

  bool get enabled => _prefs.getString('cortexCloudSyncEnabled') == '1';
  set enabled(bool v) => _prefs.saveString('cortexCloudSyncEnabled', v ? '1' : '0');

  DateTime? get lastSyncedAt {
    final v = int.tryParse(_prefs.getString('cortexCloudSyncAt'));
    return v == null ? null : DateTime.fromMillisecondsSinceEpoch(v);
  }

  Future<({bool ok, String reason})> run() async {
    if (!CortexLicense.instance.hasFeature('cloud-sync')) {
      return (ok: false, reason: 'Cloud sync is a Cortex Pro feature.');
    }
    if (!enabled) return (ok: false, reason: 'Cloud sync is turned off.');
    // TODO: encrypt + POST snapshot to `endpoint`. Stubbed for now.
    _prefs.saveString('cortexCloudSyncAt', DateTime.now().millisecondsSinceEpoch.toString());
    return (ok: true, reason: 'Sync stub: snapshot recorded locally (upload not yet implemented).');
  }
}
